x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java

```diff
@@ -100,84 +79,72 @@ public void validate(ShardFollowTask params, ClusterState clusterState) {
                                                  PersistentTasksCustomMetaData.PersistentTask<ShardFollowTask> taskInProgress,
                                                  Map<String, String> headers) {
-        return new ShardFollowNodeTask(id, type, action, getDescription(taskInProgress), parentTaskId, headers);
-    }
+        ShardFollowTask params = taskInProgress.getParams();
+        logger.info("{} Creating node task to track leader shard {}, params [{}]",
+            params.getFollowShardId(), params.getLeaderShardId(), params);
 
-    @Override
-    protected void nodeOperation(final AllocatedPersistentTask task, final ShardFollowTask params, final PersistentTaskState state) {
-        ShardFollowNodeTask shardFollowNodeTask = (ShardFollowNodeTask) task;
-        Client leaderClient = wrapClient(params.getLeaderClusterAlias() != null ?
-                this.client.getRemoteClusterClient(params.getLeaderClusterAlias()) : this.client, params);
-        Client followerClient = wrapClient(this.client, params);
-        IndexMetadataVersionChecker imdVersionChecker = new IndexMetadataVersionChecker(params.getLeaderShardId().getIndex(),
-                params.getFollowShardId().getIndex(), client, leaderClient);
-        logger.info("[{}] initial leader mapping with follower mapping syncing", params);
-        imdVersionChecker.updateMapping(1L /* Force update, version is initially 0L */, e -> {
-            if (e == null) {
-                logger.info("Starting shard following [{}]", params);
-                fetchGlobalCheckpoint(followerClient, params.getFollowShardId(),
-                        followGlobalCheckPoint -> {
-                            shardFollowNodeTask.updateProcessedGlobalCheckpoint(followGlobalCheckPoint);
-                            prepare(leaderClient, followerClient, shardFollowNodeTask, params, followGlobalCheckPoint, imdVersionChecker);
-                        }, task::markAsFailed);
-            } else {
-                shardFollowNodeTask.markAsFailed(e);
-            }
-        });
-    }
-
-    void prepare(Client leaderClient, Client followerClient, ShardFollowNodeTask task, ShardFollowTask params,
-                 long followGlobalCheckPoint,
-                 IndexMetadataVersionChecker imdVersionChecker) {
-        if (task.isRunning() == false) {
-            // TODO: need better cancellation control
-            return;
+        final Client leaderClient;
+        if (params.getLeaderClusterAlias() != null) {
+            leaderClient = wrapClient(client.getRemoteClusterClient(params.getLeaderClusterAlias()), params);
+        } else {
+            leaderClient = wrapClient(client, params);
         }
+        Client followerClient = wrapClient(client, params);
+        BiConsumer<TimeValue, Runnable> scheduler =
+            (delay, command) -> threadPool.schedule(delay, Ccr.CCR_THREAD_POOL_NAME, command);
```

Comment:
do we really need a thread pool now? can't it be generic or something? we rarely use it.

Reply:
I don't know. We are still living with an unbounded generic thread pool. I would rather see the size of this CCR thread pool shrunk significantly (it is still 32 threads) with a reasonable queue depth for these tasks than forking them off to the generic thread pool where we have no longer have control nor visibility into outstanding CCR tasks.

Reply:
everything that is doable as a separate PR should ™️